File: JavaParallelSuite/src/edu/lsu/cct/parallelsuite/bench/test/TestLocks.java
package edu.lsu.cct.parallelsuite.bench.test;

import edu.lsu.cct.parallelsuite.bench.Misc;
import edu.lsu.cct.parallelsuite.locks.*;

import java.util.LinkedList;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Supplier;

public class TestLocks {
    private static final int THREADS = 12;
    private static final int COUNT_TO = 20000;
    private static final ExecutorService pool = Executors.newFixedThreadPool(THREADS, Misc.getDaemonThreadFactory());

    private static class Counter {
        private long value = 0;
        private final Lock lock;

        public void work() {
            for (int c = 0; c < COUNT_TO; c++) {
                lock.lock();
                try {
                    ++value;
                } finally {
                    lock.unlock();
                }
            }
        }

        public boolean verify() {
            return value == (THREADS * COUNT_TO);
        }

        public Counter(Supplier<Lock> lockSupplier) {
            lock = lockSupplier.get();
        }
    }

    private static boolean testLock(Supplier<Lock> lockSupplier) {
        var counter = new Counter(lockSupplier);
        var futs = new LinkedList<CompletableFuture<?>>();

        for (int i = 0; i < THREADS; i++) {
            futs.add(CompletableFuture.runAsync(counter::work, pool));
        }

        for (var fut : futs) {
            fut.join();
        }

        var v = counter.verify();
        assert v;
        return v;
    }

    private static final int TRIES = 30;

    @SafeVarargs
    private static void testLocksRepeatedly(Supplier<Lock>... lockSuppliers) {
        for (int t = 0; t < TRIES; t++) {
            for (var lockSupplier : lockSuppliers) {
                if (!testLock(lockSupplier)) {
                    System.out.printf("A test failed! %s\n", lockSupplier.get().getClass().getName());
                    return;
                }
            }
        }

        System.out.println("All tests passed.");
    }

    public static void main(String[] args) {
        testLocksRepeatedly(
                ReentrantLock::new,
                TASLock::new,
                TTASLock::new,
                () -> new ALock(THREADS),
                BackoffLock::new,
                TwoCounterLock::new,
                CLHLock::new,
                MCSLock::new,
                IdLock::new,
                TIdLock::new
        );
    }
}

File: CppParallelSuite/bench/test/TestLocks.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <mutex>
#include "../../locks/ALock.hpp"
#include "../../locks/OptimizedALock.hpp"
#include "../../locks/IdLock.hpp"
#include "../../locks/TIdLock.hpp"
#include "../../locks/TASLock.hpp"
#include "../../locks/TTASLock.hpp"
#include "../../MutexType.hpp"
#include "../../locks/BackoffLock.hpp"
#include "../../locks/CLHLock.hpp"
#include "../../locks/MCSLock.hpp"
#include "../../locks/TwoCounterLock.hpp"

namespace parallel_test::locks {
    using namespace parallel_suite;
    using namespace parallel_suite::locks;

    constexpr static int Threads = 12;
    constexpr static int CountTo = 20000;

    template <MutexType M>
    void work(int threadId, usize& counter, M& theMutex) {
        for (int c = 0; c < CountTo; ++c) {
            std::lock_guard lock(theMutex);
            ++counter;
        }
    }

    template <MutexType M>
    bool testLock() {
        std::vector<std::thread> workers;

        usize counter = 0;
        M theMutex;

        for (int threadId = 0; threadId < Threads; ++threadId) {
            workers.emplace_back(work<M>, threadId, std::ref(counter), std::ref(theMutex));
        }

        for (auto&& worker : workers) {
            worker.join();
        }

        auto target = Threads * CountTo;

        return target == counter;
    }
} // parallel_test::locks

using namespace parallel_test::locks;

template <MutexType... Mutices>
bool testLocksRepeatedly() {
    constexpr static int Tries = 30;
    bool ok = true;

    for (int t = 0; t < Tries; ++t) {
        ok = ok && (... && testLock<Mutices>());
    }

    if (ok) {
        std::cout << "All tests passed." << std::endl;
    } else {
        std::cout << "A test failed!" << std::endl;
    }

    return ok;
}

int main() {
    bool ok = testLocksRepeatedly<
            std::mutex,
            std::recursive_mutex,
            TASLock,
            TTASLock,
            ALock<Threads>,
            OptimizedALock<Threads>,
            BackoffLock<>,
            CLHLock,
            MCSLock,
            IdLock,
            TIdLock,
            TwoCounterLock
    >();

    return ok ? 0
              : 0xBAD;
}
